ruma-appservice-api/src/event/push_events/v1.rs

```diff
@@ -36,6 +44,84 @@ impl<'a> Request<'a> {
     }
 }
 
+#[derive(Debug, Deserialize)]
+struct EventDeHelper {
+    /// the Matrix event type string "m.room.whatever".
+    #[serde(rename = "type")]
+    pub ev_type: String,
+
+    /// If `state_key` is present the event will be deserialized as a state event.
+    pub state_key: Option<IgnoredAny>,
+
+    /// If no `state_key` is found but an `event_id` is present the event
+    /// will be deserialized as a message event.
+    pub event_id: Option<IgnoredAny>,
+
+    /// If no `event_id` or `state_key` are found but a `room_id` is present
+    /// the event will be deserialized as a ephemeral event.
+    pub room_id: Option<RoomId>,
+
+    /// If this `UnsignedData` contains a `redacted_because` key the event is
+    /// immediately deserialized as a redacted event.
+    pub unsigned: Option<UnsignedDeHelper>,
+}
+
+// all observed room events go into "Joined Room" since appservice transactions
+// aren't bound to a specific user
+impl TryInto<sync_events::Response> for IncomingRequest {
+    type Error = serde_json::Error;
+
+    fn try_into(self) -> Result<sync_events::Response, Self::Error> {
```

Comment:
I would prefer if this was an inherent method with a more meaningful name and some docs, for better discoverability.